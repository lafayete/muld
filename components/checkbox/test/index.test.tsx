import * as React from 'react';
import { render, cleanup, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom';
import '@testing-library/jest-dom/extend-expect';
import 'jest-styled-components';
import Checkbox from '..';
import CheckboxGroup from '../../checkbox-group';

const { useState } = React;

const ControlledCheckbox = function () {
    const [checked, setChecked] = useState(false);
    return (
        <Checkbox
            checked={checked}
            onChange={(checked) => {
                setChecked(checked);
            }}
        ></Checkbox>
    );
};

const ControlledCheckboxGroup = function () {
    const [value, setValue] = useState<any[]>([]);
    return (
        <CheckboxGroup
            value={value}
            onChange={(checkedList) => {
                setValue(checkedList);
            }}
        >
            <Checkbox name="a">a</Checkbox>
            <Checkbox name="b">b</Checkbox>
        </CheckboxGroup>
    );
};

function expectChecked(el: Element | null) {
    expect(el).toHaveClass('mul-checkbox__icon--checked');
}

function expectNotChecked(el: Element) {
    expect(el).not.toHaveClass('mul-checkbox__icon--checked');
}

function expectToggle(clickedEl: Element, initChecked: boolean = false) {
    if (initChecked) {
        fireEvent.click(clickedEl);
        expectNotChecked(clickedEl);
        fireEvent.click(clickedEl);
        expectChecked(clickedEl);
    } else {
        fireEvent.click(clickedEl);
        expectChecked(clickedEl);
        fireEvent.click(clickedEl);
        expectNotChecked(clickedEl);
    }
}

describe('Checkbox', () => {
    afterEach(cleanup);

    it('usage of uncontrolled component and defaultChecked should be valid', () => {
        const { container } = render(<Checkbox defaultChecked={true}></Checkbox>);
        const clickedEl = container.querySelector('.mul-checkbox__icon')!;
        expectToggle(clickedEl, true);
    });

    it('usage of controlled component should be valid', () => {
        const { container: ct1 } = render(<Checkbox checked={false}></Checkbox>);
        const iconEl1 = ct1.querySelector('.mul-checkbox__icon')!;
        fireEvent.click(iconEl1);
        expectNotChecked(iconEl1);

        const { container: ct2 } = render(<ControlledCheckbox></ControlledCheckbox>);
        const clickedEl = ct2.querySelector('.mul-checkbox__icon')!;
        expectToggle(clickedEl, false);
    });

    it('disabled', () => {
        const { container } = render(<Checkbox disabled></Checkbox>);
        const clickedEl = container.querySelector('.mul-checkbox__icon')!;
        fireEvent.click(clickedEl);
        expectNotChecked(clickedEl);
    });

    it('should be uncontrolled when used within CheckboxGroup', () => {
        const { container } = render(
            <CheckboxGroup>
                <Checkbox name="a" checked={true}>
                    a
                </Checkbox>
                <Checkbox name="b">b</Checkbox>
            </CheckboxGroup>,
        );
        const clickedEls = container.querySelectorAll('.mul-checkbox__icon');
        expectNotChecked(clickedEls[0]);
    });

    it('click event should be always triggered', () => {
        const onClick = jest.fn();
        const { container } = render(<Checkbox disabled onClick={onClick}></Checkbox>);
        const clickedEl = container.querySelector('.mul-checkbox__icon')!;
        fireEvent.click(clickedEl);
        expect(onClick).toBeCalled();
    });

    it('change event should not be triggered when the box was disabled', () => {
        const onChange = jest.fn();
        const { container } = render(<Checkbox disabled onChange={onChange}></Checkbox>);
        const clickedEl = container.querySelector('.mul-checkbox__icon')!;
        fireEvent.click(clickedEl);
        expect(onChange).not.toBeCalled();
    });

    it('prop: iconSize', () => {
        const { container } = render(<Checkbox iconSize={100}></Checkbox>);
        const iconEl = container.querySelector('.mul-checkbox__icon')!;
        expect(iconEl).toMatchSnapshot();
    });

    it('prop: shape', () => {
        const { container } = render(<Checkbox shape="square"></Checkbox>);
        const el = container.querySelector('.mul-checkbox')!;
        expect(el).toMatchSnapshot();
    });

    it('prop: checkedColor', () => {
        const { container } = render(<Checkbox checkedColor="red"></Checkbox>);
        const iconEl = container.querySelector('.mul-checkbox__icon')!;
        expect(iconEl).toMatchSnapshot();
    });

    it('prop: label position', () => {
        const { container } = render(<Checkbox labelPosition="left"></Checkbox>);
        const el = container.querySelector('.mul-checkbox')!;
        expect(el).toMatchSnapshot();
    });
});

describe('CheckboxGroup', () => {
    afterEach(cleanup);
    it('should work correctly [uncontrolled component]', () => {
        const { container } = render(
            <CheckboxGroup>
                <Checkbox name="a">a</Checkbox>
                <Checkbox name="b">b</Checkbox>
            </CheckboxGroup>,
        );
        const clickedEls = container.querySelectorAll('.mul-checkbox__icon');
        expectToggle(clickedEls[0]);
        expectToggle(clickedEls[1]);
    });

    it('onChange should be called by correct argument [uncontrolled component]', () => {
        const onChange = jest.fn();
        const { container } = render(
            <CheckboxGroup onChange={onChange}>
                <Checkbox name="a">a</Checkbox>
                <Checkbox name="b">b</Checkbox>
            </CheckboxGroup>,
        );
        const clickedEls = container.querySelectorAll('.mul-checkbox__icon');
        fireEvent.click(clickedEls[0]);
        expect(onChange).toBeCalledWith(['a']);
        fireEvent.click(clickedEls[1]);
        expect(onChange).toBeCalledWith(['a', 'b']);

        fireEvent.click(clickedEls[0]);
        fireEvent.click(clickedEls[1]);
        expect(onChange).toBeCalledWith([]);
    });

    it('work correctly [controlled component]', () => {
        const { container } = render(<ControlledCheckboxGroup />);
        const clickedEls = container.querySelectorAll('.mul-checkbox__icon');
        expectToggle(clickedEls[0]);
        expectToggle(clickedEls[1]);
    });

    it('should work correctly with parameter of options', () => {
        const options = [{ name: 'a' }, { name: 'b' }];
        const { container } = render(<CheckboxGroup options={options} />);
        const clickedEls = container.querySelectorAll('.mul-checkbox__icon');
        expectToggle(clickedEls[0]);
        expectToggle(clickedEls[1]);
    });

    it('override checkbox disabled ', () => {
        const { container } = render(
            <CheckboxGroup disabled>
                <Checkbox name="a" disabled={false}>
                    a
                </Checkbox>
                <Checkbox name="b">b</Checkbox>
            </CheckboxGroup>,
        );
        const clickedEls = container.querySelectorAll('.mul-checkbox__icon');
        fireEvent.click(clickedEls[0]);
        expectNotChecked(clickedEls[0]);
    });

    it('prop: max ', () => {
        const { container } = render(
            <CheckboxGroup max={1}>
                <Checkbox name="a">a</Checkbox>
                <Checkbox name="b">b</Checkbox>
            </CheckboxGroup>,
        );
        const clickedEls = container.querySelectorAll('.mul-checkbox__icon');
        fireEvent.click(clickedEls[0]);
        fireEvent.click(clickedEls[1]);
        expectNotChecked(clickedEls[1]);
    });
});
