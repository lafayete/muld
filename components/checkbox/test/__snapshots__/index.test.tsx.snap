// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkbox prop: checkedColor 1`] = `
.c0.mul-icon {
  position: relative;
  display: inline-block;
  font-family: 'muld-icon' !important;
  font-size: inherit;
  text-rendering: auto;
  -webkit-font-smoothing: antialiased;
}

.c0.mul-icon::before {
  display: inline-block;
}

.c0.mul-icon .mul-icon__image {
  width: 1em;
  height: 1em;
  object-fit: contain;
}

.c0.mul-icon-success::before {
  content: '\\F0C8';
}

<div
  class="mul-checkbox__icon mul-checkbox__icon--round"
>
  <i
    class="c0 mul-icon mul-icon-success"
    content="F0C8"
    name="success"
    style="font-size: 0.8em;"
  />
</div>
`;

exports[`Checkbox prop: iconSize 1`] = `
.c0.mul-icon {
  position: relative;
  display: inline-block;
  font-family: 'muld-icon' !important;
  font-size: inherit;
  text-rendering: auto;
  -webkit-font-smoothing: antialiased;
}

.c0.mul-icon::before {
  display: inline-block;
}

.c0.mul-icon .mul-icon__image {
  width: 1em;
  height: 1em;
  object-fit: contain;
}

.c0.mul-icon-success::before {
  content: '\\F0C8';
}

<div
  class="mul-checkbox__icon mul-checkbox__icon--round"
  style="font-size: 100px;"
>
  <i
    class="c0 mul-icon mul-icon-success"
    content="F0C8"
    name="success"
    style="font-size: 0.8em;"
  />
</div>
`;

exports[`Checkbox prop: label position 1`] = `
.c0.mul-checkbox {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  overflow: hidden;
  cursor: pointer;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.c0.mul-checkbox--disabled {
  cursor: not-allowed;
}

.c0.mul-checkbox--horizontal {
  margin-right: 12px;
}

.c0.mul-checkbox .mul-checkbox--label-disabled {
  cursor: default;
}

.c0.mul-checkbox .mul-checkbox__icon {
  -webkit-flex: none;
  -ms-flex: none;
  flex: none;
  height: 1em;
  font-size: 20px;
  line-height: 1em;
  cursor: pointer;
}

.c0.mul-checkbox .mul-checkbox__icon .mul-icon {
  display: block;
  box-sizing: border-box;
  width: 1.25em;
  height: 1.25em;
  color: transparent;
  font-size: 0.8em;
  line-height: 1.25;
  text-align: center;
  border: 1px solid #c8c9cc;
  -webkit-transition-duration: 0.2s;
  transition-duration: 0.2s;
  -webkit-transition-property: color,border-color,background-color;
  transition-property: color,border-color,background-color;
}

.c0.mul-checkbox .mul-checkbox__icon--round .mul-icon {
  border-radius: 100%;
}

.c0.mul-checkbox .mul-checkbox__icon--checked .mul-icon {
  color: #fff;
  background-color: #108ee9;
  border-color: #108ee9;
}

.c0.mul-checkbox .mul-checkbox__icon--disabled {
  cursor: not-allowed;
}

.c0.mul-checkbox .mul-checkbox__icon--disabled .mul-icon {
  background-color: #ebedf0;
  border-color: #c8c9cc;
}

.c0.mul-checkbox .mul-checkbox__icon--disabled.mul-checkbox__icon--checked .mul-icon {
  color: #c8c9cc;
}

.c0.mul-checkbox .mul-checkbox__label {
  margin-left: 8px;
  color: #323233;
  line-height: 20px;
}

.c0.mul-checkbox .mul-checkbox__label--left {
  margin: 0 8px 0 0;
}

.c0.mul-checkbox .mul-checkbox__label--disabled {
  color: #c8c9cc;
}

.c1.mul-icon {
  position: relative;
  display: inline-block;
  font-family: 'muld-icon' !important;
  font-size: inherit;
  text-rendering: auto;
  -webkit-font-smoothing: antialiased;
}

.c1.mul-icon::before {
  display: inline-block;
}

.c1.mul-icon .mul-icon__image {
  width: 1em;
  height: 1em;
  object-fit: contain;
}

.c1.mul-icon-success::before {
  content: '\\F0C8';
}

<div
  class="c0 mul-checkbox"
>
  <span
    class="mul-checkbox__label mul-checkbox__label--left"
  />
  <div
    class="mul-checkbox__icon mul-checkbox__icon--round"
  >
    <i
      class="c1 mul-icon mul-icon-success"
      content="F0C8"
      name="success"
      style="font-size: 0.8em;"
    />
  </div>
</div>
`;

exports[`Checkbox prop: shape 1`] = `
.c0.mul-checkbox {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  overflow: hidden;
  cursor: pointer;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.c0.mul-checkbox--disabled {
  cursor: not-allowed;
}

.c0.mul-checkbox--horizontal {
  margin-right: 12px;
}

.c0.mul-checkbox .mul-checkbox--label-disabled {
  cursor: default;
}

.c0.mul-checkbox .mul-checkbox__icon {
  -webkit-flex: none;
  -ms-flex: none;
  flex: none;
  height: 1em;
  font-size: 20px;
  line-height: 1em;
  cursor: pointer;
}

.c0.mul-checkbox .mul-checkbox__icon .mul-icon {
  display: block;
  box-sizing: border-box;
  width: 1.25em;
  height: 1.25em;
  color: transparent;
  font-size: 0.8em;
  line-height: 1.25;
  text-align: center;
  border: 1px solid #c8c9cc;
  -webkit-transition-duration: 0.2s;
  transition-duration: 0.2s;
  -webkit-transition-property: color,border-color,background-color;
  transition-property: color,border-color,background-color;
}

.c0.mul-checkbox .mul-checkbox__icon--round .mul-icon {
  border-radius: 100%;
}

.c0.mul-checkbox .mul-checkbox__icon--checked .mul-icon {
  color: #fff;
  background-color: #108ee9;
  border-color: #108ee9;
}

.c0.mul-checkbox .mul-checkbox__icon--disabled {
  cursor: not-allowed;
}

.c0.mul-checkbox .mul-checkbox__icon--disabled .mul-icon {
  background-color: #ebedf0;
  border-color: #c8c9cc;
}

.c0.mul-checkbox .mul-checkbox__icon--disabled.mul-checkbox__icon--checked .mul-icon {
  color: #c8c9cc;
}

.c0.mul-checkbox .mul-checkbox__label {
  margin-left: 8px;
  color: #323233;
  line-height: 20px;
}

.c0.mul-checkbox .mul-checkbox__label--left {
  margin: 0 8px 0 0;
}

.c0.mul-checkbox .mul-checkbox__label--disabled {
  color: #c8c9cc;
}

.c1.mul-icon {
  position: relative;
  display: inline-block;
  font-family: 'muld-icon' !important;
  font-size: inherit;
  text-rendering: auto;
  -webkit-font-smoothing: antialiased;
}

.c1.mul-icon::before {
  display: inline-block;
}

.c1.mul-icon .mul-icon__image {
  width: 1em;
  height: 1em;
  object-fit: contain;
}

.c1.mul-icon-success::before {
  content: '\\F0C8';
}

<div
  class="c0 mul-checkbox"
>
  <div
    class="mul-checkbox__icon mul-checkbox__icon--square"
  >
    <i
      class="c1 mul-icon mul-icon-success"
      content="F0C8"
      name="success"
      style="font-size: 0.8em;"
    />
  </div>
  <span
    class="mul-checkbox__label mul-checkbox__label--right"
  />
</div>
`;
